### PR TITLE
Add Service Binding configuration for Kafka

### LINKED ISSUE #15631: K8s Service Binding for Kafka
Support setting up Kafka via the new Service Binding setup

- `type` should be `kafka`
- `provider` could be anything

What keys are required/supported?

### PR DESCRIPTION
**Pull request by geoand**
Fixes: #15631

### CODE REVIEW

**Comment by geoand** [comment]
@cescoffier I haven't added the rest of the Kafka properties since I couldn't find a list of them. Can you point me in the right direction please?

**Comment by cescoffier** [comment]
@geoand which property are you talking about? 

I suppose these:

* https://kafka.apache.org/documentation/#producerconfigs
* https://kafka.apache.org/documentation/#consumerconfigs

**Comment by geoand** [comment]
Yeah, those are useful! Thanks!

**Comment by geoand** [comment]
@wtrocki I don't really know what the `user` and `password` binding properties map to with regards to Kafka properties.

**Comment by geoand** [comment]
> @geoand which property are you talking about?
> 
> I suppose these:
> 
> * https://kafka.apache.org/documentation/#producerconfigs
> * https://kafka.apache.org/documentation/#consumerconfigs

I added a couple more properties

**Comment by cescoffier** [comment]
@geoand user/password are used as follows:

```
%prod.kafka.security.protocol=SASL_SSL
%prod.kafka.sasl.mechanism=PLAIN
%prod.kafka.sasl.jaas.config=org.apache.kafka.common.security.plain.PlainLoginModule required \
    username="${KAFKA_USERNAME}" \
    password="${KAFKA_PASSWORD}";
```

**Comment by geoand** [comment]
WOW.

`kafka.sasl.jaas.config` looks pretty messed up :).

I'll add that later on tonight

**Comment by geoand** [comment]
@cescoffier pushed a change based on what you explained

File: extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/runtime/KafkaBindingConverter.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+            properties.put("kafka.security.protocol", securityProtocol);
+        }
+
+        String saslMechanism = binding.getProperties().get("saslmechanism");
+        if (saslMechanism != null) {
+            properties.put("kafka.sasl.mechanism", saslMechanism);
+        }
+        String user = binding.getProperties().get("user");
+        String password = binding.getProperties().get("password");
+        if ("PLAIN".equals(saslMechanism) && (user != null) && (password != null)) {
```
Comment (geoand) [inline extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/runtime/KafkaBindingConverter.java:43]: Is this supposed to be `PLAIN` or `PLAINTEXT`?

Comment (cescoffier) [reply extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/runtime/KafkaBindingConverter.java:43]: https://docs.confluent.io/platform/current/kafka/authentication_sasl/authentication_sasl_plain.html

```
`PLAIN` versus `PLAINTEXT` Do not confuse the SASL mechanism PLAIN with no SSL encryption being called PLAINTEXT. Configuration parameters such as sasl.enabled.mechanisms or sasl.mechanism.inter.broker.protocol may be configured to use SASL mechanism PLAIN, whereas security.inter.broker.protocol or listeners may be configured for no SSL encryption SASL_PLAINTEXT.
```

Comment (wtrocki) [reply extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/runtime/KafkaBindingConverter.java:43]: This is what our operator is doing:
```
    this.saslMechanism = "PLAIN";
    this.securityProtocol = "SASL_SSL"; 
```

**Comment by geoand** [comment]
Is there anything else we need for this PR?

**Comment by wtrocki** [comment]
@geoand what would be the easiest way to test this with https://quarkus.io/guides/kafka guide?

**Comment by geoand** [comment]
@wtrocki basically in that guide, you can remove `kafka.bootstrap.servers=localhost:9092` from `application.properties`.

Then Build the application with `mvn package -DskipTests`.

After that, create the proper Kafka service binding files, say under `/bindings/some-kafka`.

Finally run the application using `QUARKUS_KUBERNETES_SERVICE_BINDING_ROOT=/bindings/some-kafka java -jar target/quarkus-app/quarkus-run.jar`

File: docs/src/main/asciidoc/deploying-to-kubernetes.adoc [hunk 2]
```diff
@@ -1198,6 +1198,7 @@ Currently, the following Quarkus extensions support this feature:
 * quarkus-jdbc-mssql
 * quarkus-jdbc-mysql
 * quarkus-jdbc-postgresql
+* quarkus-kafka-client
```
Comment (cescoffier) [inline docs/src/main/asciidoc/deploying-to-kubernetes.adoc:1201]: You may add the kafka connector from reactive messaging too (even if it's by transitivity)

File: extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaProcessor.java [hunk 3]
```diff
@@ -394,6 +395,16 @@ UnremovableBeanBuildItem ensureJsonParserAvailable() {
                 "javax.json.bind.Jsonb");
     }
 
+    @BuildStep
+    void registerServiceBinding(Capabilities capabilities,
```
Comment (cescoffier) [inline extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaProcessor.java:399]: For my curiosity is it binding or bindings.

The repo is plural: https://github.com/k8s-service-bindings/spec
The title is singular...

File: extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/runtime/KafkaBindingConverter.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+        String bootstrapServers = binding.getProperties().get("bootstrapServers");
+        if (bootstrapServers == null) {
+            bootstrapServers = binding.getProperties().get("bootstrap-servers");
+        }
+        if (bootstrapServers != null) {
+            properties.put("kafka.bootstrap.servers", bootstrapServers);
+        }
+
+        String securityProtocol = binding.getProperties().get("securityprotocol");
```
Comment (cescoffier) [inline extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/runtime/KafkaBindingConverter.java:32]: why is it boostrapServers but securityprotocol?

File: extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/runtime/KafkaBindingConverter.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+        String saslMechanism = binding.getProperties().get("saslmechanism");
+        if (saslMechanism != null) {
+            properties.put("kafka.sasl.mechanism", saslMechanism);
+        }
+        String user = binding.getProperties().get("user");
+        String password = binding.getProperties().get("password");
+        if ("PLAIN".equals(saslMechanism) && (user != null) && (password != null)) {
+            properties.put("kafka.sasl.jaas.config",
+                    String.format("org.apache.kafka.common.security.plain.PlainLoginModule required %s %s", user, password));
```
Comment (cescoffier) [inline extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/runtime/KafkaBindingConverter.java:45]: Here it's tricky. Format and module implementation are deployment-specific. By doing this, we prevent the usage of service bindings when the kafka broker use Kerberos or another type of authentication.

Comment (cescoffier) [reply extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/runtime/KafkaBindingConverter.java:45]: I believe it's fine for now, but we may need to add a hint (a file) in the service bindings.

File: extensions/kubernetes-service-binding/runtime/src/main/java/io/quarkus/kubernetes/service/binding/runtime/KubernetesServiceBindingConfigSourceProvider.java [hunk 6]
```diff
@@ -30,6 +33,7 @@ public KubernetesServiceBindingConfigSourceProvider(String bindingRoot) {
         this.serviceBindingConverters = serviceBindingConverters;
         Path p = Paths.get(bindingRoot);
         if (!Files.exists(p)) {
+            log.warn("Binding root '" + p.toAbsolutePath().toString() + "' does not exist");
```
Comment (cescoffier) [inline extensions/kubernetes-service-binding/runtime/src/main/java/io/quarkus/kubernetes/service/binding/runtime/KubernetesServiceBindingConfigSourceProvider.java:36]: I would say "Service Binding directory" , I found "Binding root" a bit out of context if you don't know the implementation details.

File: extensions/kubernetes-service-binding/runtime/src/main/java/io/quarkus/kubernetes/service/binding/runtime/ServiceBinding.java [hunk 7]
```diff
@@ -63,6 +63,7 @@ public ServiceBinding(Path bindingDirectory) {
 
     private static Map<String, String> getFilenameToContentMap(Path directory) {
         if (!Files.exists(directory) || !Files.isDirectory(directory)) {
+            log.debug("File '" + directory + "' is not a proper service binding directory so it will skipped");
```
Comment (cescoffier) [inline extensions/kubernetes-service-binding/runtime/src/main/java/io/quarkus/kubernetes/service/binding/runtime/ServiceBinding.java:66]: Is debug right here? It seems to be a deployment issue, and a warning might be more appropriate. The "ops" team would need to be aware of such an issue. no?

**Review by cescoffier — CHANGES_REQUESTED** [review]

File: extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/runtime/KafkaBindingConverter.java (hunk 4 above)
Comment (geoand) [reply extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/runtime/KafkaBindingConverter.java:32]: I just used what was mentioned in https://github.com/quarkusio/quarkus/issues/15631#issuecomment-799305457

File: extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/runtime/KafkaBindingConverter.java (hunk 5 above)
Comment (geoand) [reply extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/runtime/KafkaBindingConverter.java:45]: Yeah, this definitely need to be improved when we have more supported types

File: extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaProcessor.java (hunk 3 above)
Comment (geoand) [reply extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaProcessor.java:399]: I don't know to tell you the truth :)

File: extensions/kubernetes-service-binding/runtime/src/main/java/io/quarkus/kubernetes/service/binding/runtime/ServiceBinding.java (hunk 7 above)
Comment (geoand) [reply extensions/kubernetes-service-binding/runtime/src/main/java/io/quarkus/kubernetes/service/binding/runtime/ServiceBinding.java:66]: Good point

**Comment by geoand** [comment]
@cescoffier is there anything else you would like me to address for this first version?

**Review by cescoffier — APPROVED** [review]

**Comment by geoand** [comment]
@wtrocki  I am going to merge this in `main`, so if you would like to test, you won't need to build from the PR

**Comment by wtrocki** [comment]
Awesome. Thank you

**Comment by geoand** [comment]
YW!

**Pull request merged**